### DISCUSSION TITLE
Regrid rectilinear  to Curvilinear

### DISCUSSION THREAD

**Discussion by pramodadhikari** [NCAR/geocat-comp#367] [Q&A]
Hi,

I have a Precipitation field from PRISM datasets (for the entire United States) and want to re-grid it to match up with the WRF dataset dimension (which is only for the western US).

PRISM precipitation data: 4km with shape 621 x 1405
prism_latitude 1D array([24.083333, 24.125, 24.166667, ..., 49.833333, 49.875, 49.916667]) and 
prism_longitude 1D array([-125. , -124.958333, -124.916667, ...,  -66.583333,  -66.541667, -66.5 ]).

WRF data: 9 km with shape 340 x 270
WRF lat is 2D XLAT (from 22 to 57 degrees) and WRF lon is 2D XLONG (from -139 to -94 degrees)

I used the following code but it gave me a ValueError: axes don't match array. 
`
import numpy as np
import xarray as xr
import geocat.comp

#PRISM data

ds_rect = xr.open_dataset('prism_file.nc')
ppt_rect   =ds_rect['precip'][:]
lat1D_rect=ds_rect.lat[:]
lon1D_rect=ds_rect.lon[:].

#WRF latitude and longitude
data = xr.open_dataset("wrf_file.nc")
newlat2D_curv=data.variables["XLAT"][0,:,:]
newlon2D_curv=data.variables["XLONG"][0,:,:]

ppt_curv = geocat.comp.rgrid2rcm(lat1D_rect, lon1D_rect, ht_rect, newlat2D_curv, newlon2D_curv)

`
Thank you in advance for any help in solving this issue.
Pramod

**Comment by anissa111** [comment]
Hi! Sorry for the delayed response. I suspect some of your inputs to your `rgrid2rcm` might not be the expected dimensions. The documentation here should provide more info about the expected size: https://geocat-comp.readthedocs.io/en/latest/user_api/generated/geocat.comp.rgrid2rcm.html

If that doesn't clear things up for you, if you could give me the sizes of the input variables I could also get a better idea of where that error is coming from.

**Comment by jaredalee** [comment]
I've been dealing with a very similar problem and the same error the last couple days, but I think I found a workaround that I'd like to share, as I think it indicates a bug in the code of `rgrid2rcm` (or if not a bug, then at least incomplete/misleading documentation).

When the data variable that you'd like to regrid from a rectangular to curvilinear grid using `geocat.f2py.rgrid2rcm` has only 2 dimensions ([lat, lon]), the function throws this `ValueError: axes don't match array`. But when the data variable you pass in has 3 dimensions (e.g., [time, lat, lon]), then `rgrid2rcm` works. If you only want to regrid a single time's field, you still need to ensure that the singleton dimension is preserved. If `time` is a coordinate of your data variable, then you can select a slice of the data along the time coordinate, and including that slice selection as a list will preserve even a singleton dimension (this dataset has daily-valued data):

```
gpep_lat1d = gpep_ds['lat']
gpep_lon1d = gpep_ds['lon']
gpep_prcp = gpep_ds['prcp'].sel(time=['2021-01-26'])
wrf_lat = wrf_ds['XLAT'][0, :, :]
wrf_lon = wrf_ds['XLONG'][0, :, :]

print(gpep_lat1d.shape)
print(gpep_lon1d.shape)
print(gpep_prcp.shape)
print(wrf_lat.shape)
print(wrf_lon.shape)

gpep_prcp_regrid = rgrid2rcm(gpep_lat1d, gpep_lon1d, gpep_prcp, wrf_lat, wrf_lon)
print(gpep_prcp_regrid.shape)
```

which results in these array shapes from the `print` statements:

```
(1400,)
(2900,)
(1, 1400, 2900)
(399, 600)
(399, 600)
(1, 399, 600)
```
The [documentation](https://geocat-f2py.readthedocs.io/en/latest/autoapi/generated/geocat/f2py/index.html#geocat.f2py.rgrid2rcm) currently states that the variable `fi` is a "multi-dimensional array", but the function currently will not accept `fi` if it has only 2 dimensions; rather, it requires at least 3 dimensions (note I did not try data arrays of 4+ dimensions). This is a limitation of `rgrid2rcm` that should either be fixed or clearly highlighted in the documentation.

Jared

**Reply by erogluorhan** [reply]
Hi @jaredalee, 

Thanks very much for sharing your workaround! Honestly, we've archived the geocat-f2py package a while ago and not maintained it since then.

We'll probably not attempt at fixing this limitation you pointed out as we defer to other regridding tools in the ecosystem; however, documenting it somewhere would obviously work well for the users of geocat-f2py. 

There might be some more documentation in the homepage about this library being archived and not maintained, which can be done with the same work I believe.

We'll do it soon (#862).

**Reply by jaredalee** [reply]
Hi @erogluorhan, good to know about the geocat-f2py package being archived and not maintained anymore. However, I first tried using rgrid2rcm in my code from the geocat-comp package. However, when I did that after upgrading to the current version of geocat-comp, I got a deprecation error message telling me to use the version from geocat-f2py:

`/glade/work/jaredlee/conda-envs/my-npl-202403/lib/python3.11/site-packages/geocat/comp/deprecated.py:116: UserWarning: To use this function, install the geocat-f2py package and import as ``geocat.f2py.rgrid2rcm``.`

So with the archiving of geocat-f2py but the presumed continued development of geocat-comp, what's the future of regridding routines like rgrid2rcm and rcm2rgrid? When you say you're deferring to some of the other regridding tools in the ecosystem, do you mean in the geocat ecosystem specifically, or the wider Earth system science Python tools ecosystem more generally?

**Reply by erogluorhan** [reply]
Hi @jaredalee , I meant the tools from the broader ecosystem actually, e.g. xESMF. However, we are always happy to hear if there are any gaps in those tools and whether GeoCAT should be filling them for the community. 

Could you please elaborate on your regridding use cases, how you use GeoCAT and/or others, if there were any things that only GeoCAT's such functions were addressing your needs for, etc?

**Reply by jaredalee** [reply]
Hi @erogluorhan, I had completely forgotten about the xESMF.Regridder tool until your comment. I was able to get that to work for my needs here. Thanks.

**Comment by erogluorhan** [comment]
Glad to hear!
